### PR TITLE
Docs (`rustdoc`, `odoc`): build and publish to gh-pages (via CI)

### PR DESCRIPTION
**Pull request by W95Psp**
This PR adds build for `rustdoc` and `odoc` (ocaml's doc tool), and bundles documentation in one `index.html` ToC.
It also adds CI for deploying to GH Pages

### CODE REVIEW

File: engine/doc/dune [hunk 1]
```diff
@@ -0,0 +1,3 @@
+(documentation
+  (package hax-engine)
+  (mld_files index))
```
Comment (franziskuskiefer) [inline engine/doc/dune:3]:
```suggestion
  (mld_files index))

```

File: engine/doc/index.mld [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    {- utilities and PPXs:
+       {ul {- {!module-Hacspeclib_macro_parser}}
+           {- {!module-Ppx_functor_application}}
+           {- {!module-Ppx_generate_features}}
+           {- {!module-Ppx_inline}}
+        }
+    }
+}
+
+
```
Comment (franziskuskiefer) [inline engine/doc/index.mld:21]:
```suggestion
```

File: .github/workflows/gh_pages.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+name: Deploy documentation to GH Pages
+
+on:
+  pull_request:
+  merge_group:
```
Comment (franziskuskiefer) [inline .github/workflows/gh_pages.yml:5]: Let's only do this on main and on dispatch.

File: .github/workflows/gh_pages.yml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    if: ${{ github.event_name == 'workflow_dispatch'  || github.event_name == 'merge_group'  }}
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v3
+    - uses: DeterminateSystems/nix-installer-action@main
+    - uses: DeterminateSystems/magic-nix-cache-action@main
+    - name: Build docs
+      run: nix build .#hax.docs --out-link ./built-docs
+    - name: Deploy to gh-pages
+      uses: peaceiris/actions-gh-pages@v3
```
Comment (franziskuskiefer) [inline .github/workflows/gh_pages.yml:21]: Let's deploy by action. Then we don't need to depend on third-party actions.
See for example https://github.com/cryspen/hacl-packages/blob/main/.github/workflows/gh-pages.yml#L142-L149
(this also needs to change of the page deployment setting to action)

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
Generally lgtm with a few nits.

File: .github/workflows/gh_pages.yml (hunk 3 above)
Comment (W95Psp) [reply .github/workflows/gh_pages.yml:5]: Makes sense, thanks!

**Comment by W95Psp** [comment]
Thanks, I didn't know about those deploy actions to GH Pages, they look nice! :)

**Review by franziskuskiefer — APPROVED** [review]
Thanks

**Pull request merged**